velox/type/Conversions.h

```diff
@@ -393,18 +393,55 @@ template <bool TRUNCATE, bool LEGACY_CAST>
 struct Converter<TypeKind::VARCHAR, void, TRUNCATE, LEGACY_CAST> {
   template <typename T>
   static std::string cast(const T& val) {
-    if constexpr (std::is_same_v<T, double> || std::is_same_v<T, float>) {
-      auto stringValue = folly::to<std::string>(val);
-      if (!FLAGS_experimental_enable_legacy_cast &&
-          stringValue.find(".") == std::string::npos &&
-          isdigit(stringValue[stringValue.length() - 1])) {
-        stringValue += ".0";
-      }
-      return stringValue;
-    }
     return folly::to<std::string>(val);
   }
 
+  static std::string cast(const float& val) {
+    if constexpr (LEGACY_CAST) {
+      auto str = folly::to<std::string>(val);
+      normalizeStandardNotation(str);
+      return str;
+    }
+
+    if (FOLLY_UNLIKELY(std::isinf(val) || std::isnan(val))) {
+      return folly::to<std::string>(val);
+    } else if (
+        (val > -10'000'000 && val <= -0.001) ||
+        (val >= 0.001 && val < 10'000'000) || val == 0.0) {
+      auto str = fmt::format("{}", val);
+      normalizeStandardNotation(str);
+      return str;
+    } else {
+      // Precision of float is at most 8 significant decimal digits.
+      auto str = fmt::format("{:.7E}", val);
+      normalizeScientificNotation(str);
+      return str;
+    }
+  }
+
+  static std::string cast(const double& val) {
+    if constexpr (LEGACY_CAST) {
+      auto str = folly::to<std::string>(val);
+      normalizeStandardNotation(str);
+      return str;
+    }
+
+    if (FOLLY_UNLIKELY(std::isinf(val) || std::isnan(val))) {
+      return folly::to<std::string>(val);
+    } else if (
+        (val > -10'000'000 && val <= -0.001) ||
+        (val >= 0.001 && val < 10'000'000) || val == 0.0) {
+      auto str = fmt::format("{}", val);
+      normalizeStandardNotation(str);
+      return str;
+    } else {
+      // Precision of float is at most 17 significant decimal digits.
+      auto str = fmt::format("{:.16E}", val);
```

Comment:
Looks like the only difference between `static std::string cast(const float& val)` and `static std::string cast(const double& val)` is the number of decimal digits defined here. Maybe consider to extract the common logic into a helper function.

Reply:
Yeah, sure, I merge the two into a single place.